### PR TITLE
Add definitions for mockdate package

### PR DESCRIPTION
**Pull request by brunolm**
- [X] Make your PR against the `master` branch.
- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [X] Run `tsc` without errors.
- [X] Run `npm run lint package-name` if a `tslint.json` is present.

If adding a new definition:
- [X] The package does not provide its own types, and you can not add them.
- [X] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [X] Create it with `npm run new-package package-name`, not by basing it on an existing project.

### CODE REVIEW

File: mockdate/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{
+    "compilerOptions": {
+        "module": "commonjs",
+        "target": "es6",
+        "noImplicitAny": true,
+        "noImplicitThis": true,
+        "strictNullChecks": false,
```
Comment (andy-ms) [inline mockdate/tsconfig.json]: Make this `true`.

File: mockdate/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
+// Type definitions for mockdate 2.0
+// Project: https://github.com/boblauer/MockDate
+// Definitions by: Bruno Leonardo Michels <https://github.com/brunolm>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+declare module "mockdate" {
+    export default MockDate;
```
Comment (andy-ms) [inline mockdate/index.d.ts]: This package does not define a default export. (See [our README](https://github.com/DefinitelyTyped/DefinitelyTyped#a-package-uses-export--but-i-prefer-to-use-default-imports-can-i-change-export--to-export-default), [their README](https://github.com/boblauer/MockDate#usage)).
Instead, just use:
```ts
export as namespace MockDate;
export function set(date: { valueOf(): number } | number | string, timezoneOffset?: number): void;
export function reset(): void;
```

**Review by ghost — COMMENTED** [review]
Needs a `tslint.json` containing `{ "extends": "../tslint.json" }`.

File: mockdate/index.d.ts (hunk 2 above)
Comment (brunolm) [inline mockdate/index.d.ts]: If I do `export as namespace MockDate` it doesn't work on Node.

Would it be ok to replace `export as namespace MockDate;` with `declare module "mockdate" {` and keep the rest as you said?

Comment (brunolm) [inline mockdate/index.d.ts]: If I do `export as namespace MockDate;` it doesn't work on Node. Would it be ok to use `declare module "mockdate" {` and keep the rest as you said?

File: mockdate/tsconfig.json (hunk 1 above)
Comment (brunolm) [inline mockdate/tsconfig.json]: Done

File: mockdate/index.d.ts (hunk 2 above)
Comment (andy-ms) [inline mockdate/index.d.ts]: I use node.js and it works for me. What error are you getting?

**Comment by brunolm** [comment]
I thought tslint.json was optional. Even in the pull request template it says: `Run npm run lint package-name if a tslint.json is present.`. I think there could be a note saying when it is not required and change the tet to make it look like it is always required.

**Comment by andy-ms** [comment]
If you had done this:
> Create it with `npm run new-package package-name`, not by basing it on an existing project.

It would have added the `tslint.json` and `strictNullChecks`.

**Comment by brunolm** [comment]
Sorry I did run `new-package` script but I think I mixed up the files. I was using another lib I created definitions for in the past as example. I deleted `tslint.json` because it looked like it was optional.

And as for the `export as namespace MockDate;` I was testing with the definition file on `typings` folder. I created the folder `node_modules/@types/mockdate` and then it worked with the definition in there.

Sorry for the trouble I caused, but I'm glad I learned some stuff with it :)

**Comment by andy-ms** [comment]
I think the difference there would be that types in `node_modules/@types` automatically get included in the compilation, but if you didn't specify `typings` under [`typeRoots`](http://www.typescriptlang.org/docs/handbook/compiler-options.html), then you would need `/// <reference types="mockdate" />` to include it.

**Comment by AramZS** [comment]
@brunolm Heads up that this does not work in my current linter. I subbed in my own Type to overwrite the one I exported and it did work, this is how I did it:
```

export class MockDate {
	/**
	 * Change the Date implementation to mock a specific date.
	 * @param Date to be set as current
	 * @param timezoneOffset? The value that should be returned by new Date().getTimezoneOffset()
	 */
	set: (date: { valueOf(): number; } | number | string, timezoneOffset?: number) => void,

	/**
	 * Restore the original Date object back to the native implementation.
	 */
	reset: () => void
}
```

**Comment by andy-ms** [comment]
@AramZS That's strange, the documentation doesn't mention anything about needing to be constructed. Based on the definition you wrote there I would assume you're doing something like this?
```ts
import { MockDate } from "mockdate";
const m = new MockDate();
m.set(...);
```

**Comment by brunolm** [comment]
@AramZS see the usage example:

```
import * as mockdate from 'mockdate';

mockdate.set(new Date());
mockdate.set('2017-01-05');
mockdate.set(new Date().getTime());

mockdate.set(new Date(), 0);

mockdate.reset();
```

**Comment by AramZS** [comment]
@brunolm @andy-ms I tried removing my version of the definition and using like described, but still no go 

`yarn add @types/mockdate`

I used it exactly as suggested above and I am still getting the same error:
`mockdate.reset();`
`Property 'reset' does not exist on type 'typeof "/Users/[username]/Dev/[projectname]/src/index"'.`

I'm using VS Code and `tslint`, latest versions of each. Still no go.

**Comment by andy-ms** [comment]
@AramZS Could you provide a complete example, including the import statement?
Your choice of editor/linter should have nothing to do with this, please test using `tsc` on the command line.

**Comment by brunolm** [comment]
Just a wield guess, do you have `allowSyntheticDefaultImports` set to `true`? Try importing as

```
import mockdate from 'mockdate';
```

or maybe set it to false.

And as Andy mentioned an example of your code would be very helpful, and if you could include your `tsconfig.json` it would be awesome.

**Pull request merged**